kuma/api/v1/plus/bookmarks.py

```diff
@@ -0,0 +1,190 @@
+                "parents": parents,
+                "created": bookmark.created,
+            }
+
+    m = (page - 1) * per_page
+    n = page * per_page
+    context = {
+        "items": list(serialize_bookmarks(users_bookmarks[m:n])),
+        "metadata": {
+            "total": users_bookmarks.count(),
+            "page": page,
+            "per_page": per_page,
+        },
+    }
+    context["csrfmiddlewaretoken"] = get_token(request)
+    return JsonResponse(context)
+
+
+@never_cache
+@require_http_methods(["GET", "POST"])
+@is_subscriber
+def bookmarked(request):
+
+    url = (
+        (request.GET.get("url") or "")
+        .strip()
+        .replace("https://developer.mozilla.org", "")
+    )
+    if not url:
+        return HttpResponseBadRequest("missing 'url'")
+    if not valid_url(url):
+        return HttpResponseBadRequest("invalid 'url'")
+
+    absolute_url = f"{settings.BOOKMARKS_BASE_URL}{url}/index.json"
+
+    if request.method == "POST":
+        # Let's toggle!
+        try:
+            documenturl = DocumentURL.objects.get(uri=DocumentURL.normalize_uri(url))
+            assert not documenturl.invalid
+        except DocumentURL.DoesNotExist:
+            response = download_url(absolute_url)
+            metadata = response.json()["doc"]
+            # Because it's so big, don't store certain fields that are
+            # not helpful.
+            metadata.pop("body", None)
+            metadata.pop("toc", None)
+            metadata.pop("sidebarHTML", None)
+            metadata.pop("other_translations", None)
+            metadata.pop("flaws", None)
+
+            documenturl = DocumentURL.objects.create(
+                uri=DocumentURL.normalize_uri(url),
+                absolute_url=absolute_url,
+                metadata=metadata,
+            )
+
+        users_bookmarks = Bookmark.objects.filter(
+            user_id=request.user.id, documenturl=documenturl
+        )
+
+        for bookmark in users_bookmarks:
+            # If had a bookmark before, only need to toggle the 'deleted'
+            if bookmark.deleted:
+                bookmark.deleted = None
+
+                # If a user deletes a bookmark and then decides to bookmark it again
+                # it's because they used the "undo" functionality in the front-end.
+                # If this is the case, the most recently modified bookmark is going
+                # to be the one they deleted.
+                was_undo = False
+                for most_recently_modified in Bookmark.objects.filter(
+                    user_id=request.user.id,
+                ).order_by("-modified")[:1]:
+                    was_undo = most_recently_modified.id == bookmark.id
+                if not was_undo:
+                    # When you undo, it doesn't change the 'created' date.
+                    # But if you've re-bookmarked it after some time
+                    bookmark.created = timezone.now()
```

Comment:
Hm, I wonder if we could make this simpler. What do you think if we just compare `bookmark.deleted` with the current time, and if that difference is less than some short period of time (like an hour or less, or maybe just a few minutes or less?), we won't change `bookmark.created`, otherwise we will?

Reply:
The idea is that if you un-soft-delete something, only give it a fresh new `created` if it wasn't the last thing you soft-deleted. 

I know the current implementation is a bit complex but it works, is unit tested, and is time agnostic. 

The overall intention is that if you un-soft-delete something, let it stay in the position it was instead of going back to the most recent. Try un-bookmarking something on <HTTP://localhost:3000/en-US/plus/bookmarks> that is NOT the top-most. Then press the Undo button. It reappears where it was before. 

But, suppose you bookmark something new almost every day. Back in last month you *un-*bookmarked `/en-US/docs/Foo` and today you return to that page. When viewed, it clearly doesn't show as bookmarked in the breadcrumb bar. Now, suppose you do click to bookmark it and then immediately after you navigate to <HTTP://localhost:3000/en-US/plus/bookmarks>, now you do expect `/en-US/docs/Foo` to appear first in the list because it was the most recently bookmarked URL even if it was technically un-soft-deleted.